File: bindings/typescript/examples/outlet-invocation.ts
/**
 * Outlet invocation demo.
 *
 * Starts an in-memory relay via the caller-owned `SCP`, creates an
 * identity, creates a context with outlet capabilities, registers an outlet
 * with test vectors, mints a UCAN token, invokes the outlet with
 * authorization, and reports the result.
 *
 * Post-Phase-4 (ADR-048): all bridge operations route through an
 * explicit `SCP` instance. The free-function `mintUcan` and the static
 * `Relay.startInMemory()` factory were removed — use
 * `scp.relayStartInMemory()` and `scp.ucanMint(...)`.
 *
 * Run: bun run examples/outlet-invocation.ts
 */

import { SCP, type OutletDefinition, type UcanToken, defineOutletDefinition } from "../src/index";

async function main(): Promise<void> {
  const scp = new SCP({ storage: { type: "in_memory" } });
  let relay: Awaited<ReturnType<SCP["relayStartInMemory"]>> | null = null;
  try {
    // 1. Start an in-memory relay.
    relay = await scp.relayStartInMemory();
    console.log(`Relay listening at ${relay.relayUrl}`);

    // 2. Create an identity.
    const identity = await scp.identityCreate("in_memory");
    console.log(`Identity DID: ${identity.did}`);

    // 3. Wire the bridge's transport to the relay.
    await scp.configureRelayTransport(relay.relayUrl, identity.did);

    // 4. Define an outlet with test vectors.
    const weatherOutlet: OutletDefinition = defineOutletDefinition({
      name: "weather",
      description: "Get current weather for a city",
      inputSchema: {
        type: "object",
        properties: { city: { type: "string" } },
        required: ["city"],
      },
      outputSchema: {
        type: "object",
        properties: {
          tempC: { type: "number" },
          condition: { type: "string" },
        },
      },
      operator: identity.did,
      testVectors: [
        {
          input: { city: "Berlin" },
          expectedOutput: { tempC: 18, condition: "cloudy" },
          description: "Berlin weather lookup",
        },
      ],
    });

    // 5. Create a context with outlet capabilities.
    const ctx = await scp.contextCreate(
      identity,
      JSON.stringify({
        ceiling: ["messages:read", "messages:write", "outlet:call:*", "outlet:register"],
        memoryScope: "ephemeral",
        governance: "single_admin",
      }),
    );
    console.log(`Context created: ${ctx.contextId}`);

    // 6. Register the outlet.
    const outletId = await scp.outletRegister(ctx._rawHandle, weatherOutlet);
    console.log(`Registered outlet: ${outletId}`);

    // 7. Mint a UCAN token for outlet invocation.
    const ucan = (await scp.ucanMint(ctx._rawHandle, identity.did, ["outlet_call:*"])) as UcanToken;
    console.log(`UCAN minted: ${ucan.id}`);

    // 8. Invoke the outlet with the UCAN token.
    try {
      const result = await scp.outletInvoke(
        ctx._rawHandle,
        "weather",
        JSON.stringify({ city: "Berlin" }),
        identity.did,
        ucan.encoded,
      );
      console.log("Weather result:", result);
    } catch (err) {
      console.log("Outlet invocation result:", err);
    }

    // 9. Cleanup.
    await scp.contextClose(ctx._rawHandle, identity.did);
    console.log("Context closed");
  } finally {
    if (relay !== null) {
      await relay.shutdown();
    }
    await scp.shutdown(5);
    console.log("Demo complete");
  }
}

main().catch((error: unknown) => {
  console.error("Demo failed:", error);
  process.exit(1);
});

File: bindings/python/examples/outlet_invocation.py
"""Outlet invocation demo.

Starts an in-memory relay, creates an identity, creates a context with
outlet capabilities, mints a UCAN token for authorization, invokes the
outlet, and verifies the result.

Phase 4 PR 5 (#1549) moved every operation onto :class:`SCP` — see
:meth:`SCP.relay_start_in_memory`, :meth:`SCP.transport_connect`,
:meth:`SCP.identity_create`, :meth:`SCP.context_create`,
:meth:`SCP.ucan_mint`, :meth:`SCP.outlet_invoke`, and
:meth:`SCP.context_close`.

Requires a built native extension (``maturin develop --release``).
"""

import asyncio

from scp_sdk import SCP
from scp_sdk.types import Capability, CustodyType, MemoryScope


async def main() -> None:
    with SCP(storage={"type": "in_memory"}) as scp:
        # 1. Start an in-memory relay.
        relay = await scp.relay_start_in_memory()
        print(f"Relay listening at {relay.relay_url}")

        # 1b. Connect transport to the relay.
        await scp.transport_connect(relay.relay_url)

        # 2. Create an identity.
        identity = await scp.identity_create(CustodyType.IN_MEMORY)
        print(f"Identity DID: {identity.did}")

        # 3. Create a context with outlet capabilities.
        ctx = await scp.context_create(
            identity.did,
            {
                "ceiling": [
                    Capability.MESSAGES_READ.value,
                    Capability.MESSAGES_WRITE.value,
                    Capability.TOOL_INVOKE_ALL.value,
                    Capability.TOOL_REGISTER.value,
                ],
                "memory_scope": MemoryScope.EPHEMERAL.value,
                "governance": "single_admin",
            },
        )
        print(f"Context created: {ctx.context_id}")

        try:
            # 4. Mint a UCAN token authorizing outlet invocation.
            ucan_token = await scp.ucan_mint(
                ctx.context_id,
                identity.did,
                ["tool:invoke:*"],
            )
            print(f"UCAN minted: {ucan_token.token_id}")

            # 5. Invoke the outlet (requires a UCAN token).
            try:
                result = await scp.outlet_invoke(
                    ctx.context_id,
                    "weather",
                    {"city": "Berlin"},
                    identity.did,
                    ucan_token.token_id,
                )
                print(f"Weather result: {result}")
            except Exception as exc:
                # Outlet invocation may fail without a registered outlet handler.
                print(f"Outlet invocation result: {exc}")

            # 6. Cleanup.
            await scp.context_close(ctx._raw_handle, identity.did)
            print("Context closed")
        finally:
            await relay.shutdown()

    print("Demo complete")


if __name__ == "__main__":
    asyncio.run(main())
